### PR TITLE
ItemFormatter for printing items, borg list redo.

### PR DESCRIPTION
**Pull request by enkore**
Current state:

- compatible except for {formatkeys}, which is gone
  - available keys now right in --help, with additional per-key help if deemed necessary
- supports -e, exclude-from, paths like the others
  (code should/will be shared, see feature-diff branch)
- supports guaranteed_available hashlib hashes
  (to avoid varying functionality depending on environment)
  (also, the other hashes are really obscure, like MD-4)

The two profiles [formerly linked here](https://github.com/borgbackup/borg/issues/735#issuecomment-196355903) were actually a mistake ; I wasn't running the second command inside the current environment and hence still used the old version (which is the reason why these are so similar -- should've been suspicious). The first approach was actually a **lot** slower, due to the many object creations (62 % of total runtime spent inside `builtins.__build_class__`!).

I made another approach, which is more stateful and does not use the class-wrapping approach as before. Together with some optimization of IO, mainly turning line_buffering off, this version and the version from `master` are indistinguishable, +-2 % over multiple runs, and this time I made sure I'm in the right envs ;-)

The overhead of adding additional "lazy" (now it's actually "on demand") keys is very low in this approach.

### CODE REVIEW

File: borg/archiver.py [hunk 1]
```diff
@@ -1458,10 +1419,8 @@ def setup_signal_handlers():  # pragma: no cover
 
 
 def main():  # pragma: no cover
-    # Make sure stdout and stderr have errors='replace') to avoid unicode
-    # issues when print()-ing unicode file names
-    sys.stdout = io.TextIOWrapper(sys.stdout.buffer, sys.stdout.encoding, 'replace', line_buffering=True)
-    sys.stderr = io.TextIOWrapper(sys.stderr.buffer, sys.stderr.encoding, 'replace', line_buffering=True)
+    sys.stdout = io.TextIOWrapper(sys.stdout.buffer, sys.stdout.encoding, 'surrogateescape', line_buffering=True)
+    sys.stderr = io.TextIOWrapper(sys.stderr.buffer, sys.stderr.encoding, 'surrogateescape', line_buffering=True)
```
Comment (enkore) [inline borg/archiver.py]: I don't think this is necessary anymore (at all) - maybe for stderr, but I think all places use remove_surrogates when printing to stdout.

File: borg/archiver.py [hunk 2]
```diff
@@ -1140,15 +1093,22 @@ def build_parser(self, args=None, prog=None):
         subparser.add_argument('--short', dest='short',
                                action='store_true', default=False,
                                help='only print file/directory names, nothing else')
-        subparser.add_argument('--list-format', dest='listformat', type=str,
-                               help="""specify format for archive file listing
-                                (default: "{mode} {user:6} {group:6} {size:8d} {isomtime} {path}{extra}{NEWLINE}")
-                                Special "{formatkeys}" exists to list available keys""")
+        subparser.add_argument('-f', '--format', '--list-format', dest='format', type=str,
```
Comment (ThomasWaldmann) [inline borg/archiver.py]: be careful with short options, they easily collide or have different meanings at different places.
also, we might want to use -f for something else in future, like --force?

File: borg/archiver.py [hunk 3]
```diff
@@ -1,9 +1,10 @@
 from binascii import hexlify, unhexlify
 from datetime import datetime
-from hashlib import sha256
 from operator import attrgetter
+from string import Formatter
```
Comment (ThomasWaldmann) [inline borg/archiver.py]: unused import?

File: borg/helpers.py [hunk 4]
```diff
@@ -546,7 +548,7 @@ def format_line(format, data):
     except (KeyError, ValueError) as e:
         # this should catch format errors
         print('Error in lineformat: "{}" - reason "{}"'.format(format, str(e)))
-    except:
+    except Exception as e:
```
Comment (ThomasWaldmann) [inline borg/helpers.py]: oops? looks like there was a bug?

File: borg/helpers.py [hunk 5]
```diff
@@ -1080,3 +1082,111 @@ def log_multi(*msgs, level=logging.INFO):
...
+        formatter = cls(FakeArchive, "")
+        keys = []
+        keys.extend(formatter.call_keys.keys())
+        keys.extend(formatter.get_item_data({
+            b'mode': 0, b'path': '', b'user': '', b'group': '', b'mtime': 0,
+            b'uid': 0, b'gid': 0,
+        }).keys())
+        keys.sort()
+        keys.sort(key=str.isupper)
+        return keys
```
Comment (ThomasWaldmann) [inline borg/helpers.py]: return sorted(keys, ...)

File: borg/archiver.py (hunk 2 above)
Comment (enkore) [inline borg/archiver.py]: Agreed. --format is short enough anyway, --list-format is kinda "strange" however (*what other format would it be?*)

File: borg/archiver.py (hunk 3 above)
Comment (enkore) [inline borg/archiver.py]: Yep, remnant of moving the ItemFormatter class from archiver.Archiver to helpers.

File: borg/helpers.py [hunk 6]
```diff
@@ -1080,3 +1082,111 @@ def log_multi(*msgs, level=logging.INFO):
...
+            'isomtime': partial(self.format_time, b'mtime'),
+            'isoctime': partial(self.format_time, b'ctime'),
+            'isoatime': partial(self.format_time, b'atime'),
+            'mtime': partial(self.time, b'mtime'),
+            'ctime': partial(self.time, b'ctime'),
+            'atime': partial(self.time, b'atime'),
+        }
+        for hash_function in hashlib.algorithms_guaranteed:
+            self.call_keys[hash_function] = partial(self.hash_item, hash_function)
+        self.keys = set(self.call_keys) & self.format_keys
```
Comment (ThomasWaldmann) [inline borg/helpers.py]: intersection? if you wanted union, that is |.

Comment (enkore) [inline borg/helpers.py]: Intersection is what we want:  call_keys are alle keys calculated from callables, format_keys are those used by the format string. The intersecting set are those keys used by the format string which are contained in call_keys and need to be sourced from there.

Variable naming should be improved :D

File: borg/helpers.py [hunk 7]
```diff
@@ -1080,3 +1082,111 @@ def log_multi(*msgs, level=logging.INFO):
...
+            'gid': item[b'gid'],
+
+            'path': remove_surrogates(item[b'path']),
+            'bpath': item[b'path'],
+            'source': source,
+            'linktarget': source,
+            'extra': extra
+        })
+        for key in self.keys:
+            item_data[key] = self.call_keys[key](item)
```
Comment (enkore) [inline borg/helpers.py]: Note that there is no lazyness here: all requested keys are calculated, unconditionally. The "lazyness" is that we check beforehand which keys need calculating.

File: borg/helpers.py [hunk 8]
```diff
@@ -1080,3 +1082,111 @@ def log_multi(*msgs, level=logging.INFO):
...
+
+    def format_item(self, item):
+        return self.format.format(**self.get_item_data(item))
+
+    def calculate_size(self, item):
+        if stat.S_ISREG(item[b'mode']):
+            try:
+                return sum(size for _, size, _ in item[b'chunks'])
+            except KeyError:
+                pass
```
Comment (ThomasWaldmann) [inline borg/helpers.py]: how about:

return sum(chunk[1] for chunk in item.get(b'chunks', []))

The isreg check is maybe rather harmful, borg can backup the contents of devices and then isreg might be false, but chunks still in there.

File: borg/helpers.py [hunk 9]
```diff
@@ -1080,3 +1082,111 @@ def log_multi(*msgs, level=logging.INFO):
...
+    def hash_item(self, hash_function, item):
+        if b'chunks' not in item:
+            return ""
+        hash = hashlib.new(hash_function)
+        for chunk in self.archive.pipeline.fetch_many([c[0] for c in item[b'chunks']]):
+            hash.update(chunk)
+        return hash.hexdigest()
+
+    def format_time(self, key, item):
+        return format_time(safe_timestamp(item[key]))
```
Comment (ThomasWaldmann) [inline borg/helpers.py]: ctime and atime is not aways there, should use mtime if not.

File: borg/helpers.py [hunk 10]
```diff
@@ -1080,3 +1082,111 @@ def log_multi(*msgs, level=logging.INFO):
...
+
+    def time(self, key, item):
+        return safe_timestamp(item[key])
```
Comment (ThomasWaldmann) [inline borg/helpers.py]: same here.

File: borg/helpers.py (hunk 6 above)
Comment (ThomasWaldmann) [inline borg/helpers.py]: ah, right, misread that at first.

File: borg/helpers.py (hunk 9 above)
Comment (enkore) [inline borg/helpers.py]: Seems like there was a little bit "lost in translation" :)

I'm going with `timestamp_handling_fn(item.get(key) or item[b'mtime]))` to get eval-once behaviour most of the time. `item.get(key, item[b'mtime])` would be more conventional, but these pieces are literally called a gazillion times...

File: borg/helpers.py [hunk 11]
```diff
@@ -1080,3 +1082,111 @@ def log_multi(*msgs, level=logging.INFO):
...
+        class FakeArchive:
+            name = ""
+
+        formatter = cls(FakeArchive, "")
+        keys = []
+        keys.extend(formatter.call_keys.keys())
+        keys.extend(formatter.get_item_data({
+            b'mode': 0, b'path': '', b'user': '', b'group': '', b'mtime': 0,
+            b'uid': 0, b'gid': 0,
+        }).keys())
```
Comment (ThomasWaldmann) [inline borg/helpers.py]: fake_item = {...}
keys = formatter.call_keys.keys() + formatter.get_item_data(fake_item).keys()

Comment (enkore) [inline borg/helpers.py]: I don't think the dict-view types are addable.

**Comment by ThomasWaldmann** [comment]
hmm, travis seems unhappy.

but overall great change - using this we can implement a lot of stuff by just extending the available format keys (like e.g. file compressed size, compression ratio, (un)compressed unique chunks size, count of unique chunks, ...

File: borg/helpers.py (hunk 11 above)
Comment (ThomasWaldmann) [inline borg/helpers.py]: oh, I was still thinking in lists, like in python2. yes in py3, that does not work.

that would work, not sure if it helps with readability:

fake_item = {...}
keys = set(formatter.call_keys) | set(formatter.get_item_data(fake_item))

File: borg/helpers.py (hunk 9 above)
Comment (ThomasWaldmann) [inline borg/helpers.py]: if the timestamp is 0, that does not work.

File: borg/helpers.py [hunk 12]
```diff
@@ -1107,9 +1114,17 @@ class FakeArchive:
             b'mode': 0, b'path': '', b'user': '', b'group': '', b'mtime': 0,
             b'uid': 0, b'gid': 0,
         }).keys())
-        keys.sort()
-        keys.sort(key=str.isupper)
-        return keys
+        return sorted(sorted(keys), key=str.isupper)
```
Comment (ThomasWaldmann) [inline borg/helpers.py]: why do you double-sort? is the inner sort doing anything that affects the outer result?

File: borg/archiver.py [hunk 13]
```diff
@@ -438,78 +438,31 @@ def do_list(self, args):
...
-                item_data['formatkeys'] = list(item_data.keys())
-
-                print(format_line(list_format, item_data), end='')
-
+                format = "{mode} {user:6} {group:6} {size:8} {isomtime} {path}{extra}{LF}"
+            formatter = ItemFormatter(archive, format)
+            if not hasattr(sys.stdout, 'buffer'):
+                # This is a shim for supporting unit tests replacing sys.stdout with e.g. StringIO,
+                # which doesn't have an underlying buffer (= lower file object).
+                write = lambda bytestring: sys.stdout.write(bytestring.decode('utf-8', errors='replace'))
```
Comment (ThomasWaldmann) [inline borg/archiver.py]: there was no major change in the testsuite, so how does it come we suddenly need this?

File: borg/helpers.py (hunk 12 above)
Comment (enkore) [inline borg/helpers.py]: first alphabetical sort, then separate keys only composed of uppercase letters to the bottom. This way the "formatting" aids (TAB, NEWLINE, LF, CR, NUL) appear together at the bottom, not mixed between content keys.

File: borg/archiver.py (hunk 13 above)
Comment (enkore) [inline borg/archiver.py]: The new code uses sys.stdout.buffer (which is always a binary file) for two reasons

1. to have errors=surrogateescape, so we can print binary strings (=> `{bpath}`). The default stdout we create in main() has the errors=replace policy, which prohibits this.
2. it circumvents the line buffering (…actually looks like "line-fsyncing" for writes) of the stdout TextIOWrapper. This makes IO to stdout significantly faster: with approx identical total run-time `print` in the old code took 12 % of the total run time, while `write` through this object only takes 1.6 % of the total run time.

This is however incompatible with StringIO, which doesn't have the underlying object as a `buffer` attribute, because it isn't a IOWrapper.

File: borg/helpers.py (hunk 12 above)
Comment (ThomasWaldmann) [inline borg/helpers.py]: ah, interesting. i somehow had expected that the outer sort might disturb the inner sort result.

**Comment by enkore** [comment]
master branch callmap:

borg list master

this patch (yesterday, today should be slightly better):

borg list lazy

File: borg/helpers.py (hunk 12 above)
Comment (enkore) [inline borg/helpers.py]: sort is stable, but I just remembered that tuples also work for keys, so that's simpler. Will include that in the next patch rev.

File: borg/testsuite/archiver.py [hunk 14]
```diff
@@ -898,7 +898,7 @@ def test_list_list_format(self):
         self.cmd('create', test_archive, src_dir)
         output_1 = self.cmd('list', test_archive)
         output_2 = self.cmd('list', '--list-format', '{mode} {user:6} {group:6} {size:8d} {isomtime} {path}{extra}{NEWLINE}', test_archive)
-        output_3 = self.cmd('list', '--list-format', '{mtime:%s} {path}{NL}', test_archive)
+        output_3 = self.cmd('list', '--list-format', '{mtime:%s} {path}{NEWLINE}', test_archive)
```
Comment (enkore) [inline borg/testsuite/archiver.py]: While I still don't know where NL came from in the old code I'm just going to revert this and add NL as an alias for NEWLINE. No one wants to type 10 characters for a new line anyway.

File: borg/archiver.py (hunk 13 above)
Comment (ThomasWaldmann) [inline borg/archiver.py]: it somehow seems inappropriate to have a different way of doing output just for borg list.

Comment (enkore) [inline borg/archiver.py]: I think it's ok, since line buffering for list doesn't make a lot of sense, but for the other commands it does. The errors-policy is another thing, it's probably okay in most other cases to remove non-text chars from paths, but I think for list we should be able to [reliably] support what posix demands.

**Comment by enkore** [comment]
Current performance

Test case: an archive of my /home here, about 1.5 million files. List goes straight into /dev/null. FS caches are hot (2 runs each to warm them up, then average across another three runs).

master: 1m17s (1m15s .. 1m21s)
93a6515: 1m11s +- 1 s

From profiling of this test case I'd say this version is faster for two reasons
* IO / writes to stdout is three times faster (3.5 % vs. 10 % of run time)
* *Slightly* less per-item overhead: no large objects are created per item except the item itself, and we use format_map instead of format, which is about 2 % faster (I made a synthetic benchmark which were closely matches this figure from the profile data).
* No free lunch: the extra calls for size seem to add a little overhead to that block

So in summary, we're getting a little faster, but most importantly, we're *not* slower.

File: borg/helpers.py [hunk 15]
```diff
@@ -1080,3 +1082,146 @@ def log_multi(*msgs, level=logging.INFO):
...
+        'LF': '\n',
+        'SPACE': ' ',
+        'TAB': '\t',
+        'CR': '\r',
+        'NUL': '\0',
+        'NEWLINE': os.linesep,
+        'NL': os.linesep,
+    }
+    KEY_DESCRIPTIONS = {
+        'NL': 'alias of NEWLINE (OS dependent line separator)',
```
Comment (ThomasWaldmann) [inline borg/helpers.py]: you say that NL is an alias of NEWLINE, but you don't have a description for NEWLINE.

File: borg/helpers.py [hunk 16]
```diff
@@ -1080,3 +1082,146 @@ def log_multi(*msgs, level=logging.INFO):
...
+            'ctime': partial(self.time, b'ctime'),
+            'atime': partial(self.time, b'atime'),
+        }
+        for hash_function in hashlib.algorithms_guaranteed:
+            self.add_key(hash_function, partial(self.hash_item, hash_function))
+        self.used_call_keys = set(self.call_keys) & self.format_keys
+        self.item_data = static_keys
+
+    def partial_format(self, format, mapping):
+        # str.format wishlist: leave_unknown_keys_alone_and_dont_throw_up=True
```
Comment (ThomasWaldmann) [inline borg/helpers.py]: maybe rather document what this does

Comment (enkore) [inline borg/helpers.py]: I'll also add a few tests for it, since it's non-trivial

Comment (ThomasWaldmann) [inline borg/helpers.py]: i just thought how it will behave for keys that are substrings of other keys...

File: borg/helpers.py [hunk 17]
```diff
@@ -540,6 +540,16 @@ def dir_is_tagged(path, exclude_caches, exclude_if_present):
     return tag_paths
 
 
+def partial_format(format, mapping):
+    """Apply format.format_map(mapping) while preserving unknown keys"""
+    for key, value in mapping.items():
+        def a(x): print(x.groups()); return ""
```
Comment (ThomasWaldmann) [inline borg/helpers.py]: hm? left over debug code?

Comment (enkore) [inline borg/helpers.py]: Yep.

**Comment by enkore** [comment]
Some of these commits were breaking, so I will squash them when review is finished.

File: borg/archiver.py [hunk 18]
```diff
@@ -1332,15 +1286,22 @@ def build_parser(self, args=None, prog=None):
         subparser.add_argument('--short', dest='short',
                                action='store_true', default=False,
                                help='only print file/directory names, nothing else')
-        subparser.add_argument('--list-format', dest='listformat', type=str,
-                               help="""specify format for archive file listing
-                                (default: "{mode} {user:6} {group:6} {size:8d} {isomtime} {path}{extra}{NEWLINE}")
-                                Special "{formatkeys}" exists to list available keys""")
+        subparser.add_argument('--format', '--list-format', dest='format', type=str,
```
Comment (ThomasWaldmann) [inline borg/archiver.py]: maybe this should use the cmdline option deprecation mechanism, so we can get rid of the --list-format later.
grep for "deprec" in this file.

**Comment by ThomasWaldmann** [comment]
aside from deprecation handling and the squashing you wanted to do, i think this is ready.

so tell if you are ready with it, so I'll merge it.

**Comment by enkore** [comment]
The deprecation mechanism prints to stdout:
```
    def preprocess_args(self, args):
        deprecations = [
            # ('--old', '--new', 'Warning: "--old" has been deprecated. Use "--new" instead.'),
            ('--list-format', '--format', 'Warning: "--list-format" has been deprecated. Use "--format" instead.'),
        ]
        for i, arg in enumerate(args[:]):
            for old_name, new_name, warning in deprecations:
                if arg.startswith(old_name):
                    args[i] = arg.replace(old_name, new_name)
                    print(warning)
        return args
```

Should I change that to self.print_warning? Seems more logical to me to not mix command output on stdout with these warnings.

**Comment by ThomasWaldmann** [comment]
yes, call print_warning. this will also set warning return code, but i guess that'll help fixing deprecated stuff in scripts.

**Comment by enkore** [comment]
Will do.

The Travis CI fail for 3417906 looks strange, maybe a flaky / one time error?

```
    def test_list_list_format(self):
        self.cmd('init', self.repository_location)
        test_archive = self.repository_location + '::test'
        self.cmd('create', test_archive, src_dir)
        output = self.cmd('list', '--list-format', '-', test_archive)
        assert output.startswith('Warning: "--list-format" has been deprecated')
        output_1 = self.cmd('list', test_archive)
        output_2 = self.cmd('list', '--format', '{mode} {user:6} {group:6} {size:8d} {isomtime} {path}{extra}{NEWLINE}', test_archive)
        output_3 = self.cmd('list', '--format', '{mtime:%s} {path}{NL}', test_archive)
>       self.assertEqual(output_1, output_2)
E       AssertionError: 'drwxr-xr-x root   staff         0 Thu, 20[16944 chars]py\n' != 'Exception ignored in: <bound method Repos[17732 chars]py\n'
E       Diff is 17887 characters long. Set self.maxDiff to None to see it.
```

I guess where py.test cut off an exception in some Repository method occured.

**Comment by ThomasWaldmann** [comment]
I restarted the failing test run, let's see if it fails again.

**Comment by enkore** [comment]
I saw it once here now, as well. It's a LockFailed exception downwards of Repository.__del__():

```
Exception ignored in: <bound method Repository.__del__ of <Repository /tmp/tmpjo24yn28/repository>>
Traceback (most recent call last):
  File "/home/mabe/Projekte/_oss/borg/borg/repository.py", line 67, in __del__
    self.close()
  File "/home/mabe/Projekte/_oss/borg/borg/repository.py", line 159, in close
    self.lock.release()
  File "/home/mabe/Projekte/_oss/borg/borg/locking.py", line 293, in release
    with self._lock:
  File "/home/mabe/Projekte/_oss/borg/borg/locking.py", line 117, in __enter__
    return self.acquire()
  File "/home/mabe/Projekte/_oss/borg/borg/locking.py", line 140, in acquire
    raise LockFailed(self.path, str(err)) from None
borg.locking.LockFailed: ('/tmp/tmpjo24yn28/repository/lock.exclusive', "[Errno 2] No such file or directory: '/tmp/tmpjo24yn28/repository/lock.exclusive'")
```

**Comment by ThomasWaldmann** [comment]
yes, there is already a ticket about further locking improvements (which will likely be done by making repository a context manager).

are you finished or do you still need to squash things?

**Comment by enkore** [comment]
I'm done, good to merge, when Travis gives the word.
